### PR TITLE
Feat/condo/doma 8513/meters sbbol registry

### PR DESCRIPTION
**Pull request by AleX83Xpert**
The meters registry uploading via new mutation. For now we support two formats: doma-xls and sbbol.

To-do list:
- [x] add SbbolMetersImporter
- [x] add tests for AbstractMetersImporter implementations

### CODE REVIEW

File: apps/condo/domains/common/components/EmptyListContent.tsx [hunk 1]
```diff
@@ -35,12 +35,15 @@ type IEmptyListWithoutImportProps = IBaseEmptyListProps & {
     createLabel?: string
 }
 
+type TExternalImportWrapper = Pick<IImportWrapperProps, 'onFinish' | 'uploadButtonLabel' | 'importCardButton'>
```
Comment (sitozzz) [inline apps/condo/domains/common/components/EmptyListContent.tsx:38]: For better compatibility and understanding
```suggestion
type ExternalImportWrapperType = Pick<IImportWrapperProps, 'onFinish' | 'uploadButtonLabel' | 'importCardButton'>
```

File: apps/condo/domains/common/components/Import/BaseImportWrapper.tsx [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+                            .replace('{success}', successRowsRef.current)
+                            .replace('{total}', totalRowsRef.current)
+                            .replace('{genitive}', ImportGenitiveMessage.toLowerCase())
+                    }
+                    description={PartlyDataLoadedModalAlertDescription}
+                />
+            </Modal>
+            <Modal
+                title={ErrorModalTitle}
+                open={activeModal === 'error'}
```
Comment (sitozzz) [inline apps/condo/domains/common/components/Import/BaseImportWrapper.tsx:245]: Maybe move all activemodals to enum for ease of reading

File: apps/condo/domains/meter/utils/metersImporters/SbbolMetersImporter.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            const [meterType, meterNumber, meterName, nextVerificationDate, meterValue, date, additionalInfo] = nextBlock
+
+            if (meterType && meterNumber && meterValue && date) {
+                let values, meterResourceId, numberOfTariffs = 1
+
+                /**
+                 * @see https://vgkh.ru/faq-1s-zhkh/obmen-sberbank/obmen-s-sistemoj-sberbank-biznes-onlajn/#shablon-pokazaniya-schetchikov
+                 * */
+                switch (meterType) {
+                    case '1': // Газ
```
Comment (sitozzz) [inline apps/condo/domains/meter/utils/metersImporters/SbbolMetersImporter.ts:35]: replace with translation for comments

File: apps/condo/domains/meter/utils/metersImporters/SbbolMetersImporter.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        const result = {}
+        let nextBlock = metersBlocks.splice(0, METERS_BLOCK_LENGTH)
+        while (nextBlock.length === METERS_BLOCK_LENGTH) {
+            const [meterType, meterNumber, meterName, nextVerificationDate, meterValue, date, additionalInfo] = nextBlock
+
+            if (meterType && meterNumber && meterValue && date) {
+                let values, meterResourceId, numberOfTariffs = 1
+
+                /**
+                 * @see https://vgkh.ru/faq-1s-zhkh/obmen-sberbank/obmen-s-sistemoj-sberbank-biznes-onlajn/#shablon-pokazaniya-schetchikov
```
Comment (sitozzz) [inline apps/condo/domains/meter/utils/metersImporters/SbbolMetersImporter.ts:32]: i'm afraid of this link. maybe place it inside our documentation?

File: apps/condo/domains/property/utils/testSchema/factories.js [hunk 5]
```diff
@@ -96,7 +97,65 @@ const buildFakeAddressAndMeta = (withFlat, addressMetaExtraAttrs = {}) => {
...
+            floors: map(Array(parkingFloors), (...[, floor]) => {
+                const floorPlus1 = floor + 1
+                return {
+                    id: String(floorPlus1),
+                    type: 'floor',
+                    index: floorPlus1,
+                    name: String(floorPlus1),
+                    units: map(Array(unitsOnParkingFloor), (...[, n]) => ({
+                        id: String(floor * unitsOnParkingFloor + n + 1),
+                        type: 'unit',
```
Comment (sitozzz) [inline apps/condo/domains/property/utils/testSchema/factories.js:145]: You can reuse unittype from constants

File: apps/condo/domains/property/utils/testSchema/factories.js [hunk 6]
```diff
@@ -96,7 +97,65 @@ const buildFakeAddressAndMeta = (withFlat, addressMetaExtraAttrs = {}) => {
...
+            floors: map(Array(floors), (...[, floor]) => {
+                const floorPlus1 = floor + 1
+                return {
+                    id: String(floorPlus1),
+                    type: 'floor',
+                    index: floorPlus1,
+                    name: String(floorPlus1),
+                    units: map(Array(unitsOnFloor), (...[, n]) => ({
+                        id: String(floor * unitsOnFloor + n + 1),
+                        type: 'unit',
```
Comment (sitozzz) [inline apps/condo/domains/property/utils/testSchema/factories.js:119]: please use types from constants

File: apps/condo/domains/property/utils/testSchema/factories.js [hunk 7]
```diff
@@ -96,7 +97,65 @@ const buildFakeAddressAndMeta = (withFlat, addressMetaExtraAttrs = {}) => {
     return { address, addressMeta }
 }
 
+const buildPropertyMap = ({ floors = 3, unitsOnFloor = 4, parkingFloors = 2, unitsOnParkingFloor = 8 } = {}) => ({
```
Comment (sitozzz) [inline apps/condo/domains/property/utils/testSchema/factories.js:100]: You can also build n sections here

File: apps/condo/domains/common/components/EmptyListContent.tsx (hunk 1 above)
Comment (AleX83Xpert) [reply apps/condo/domains/common/components/EmptyListContent.tsx:38]: The **T** in the beginning means the same :)

File: apps/condo/domains/common/components/Import/BaseImportWrapper.tsx (hunk 2 above)
Comment (AleX83Xpert) [reply apps/condo/domains/common/components/Import/BaseImportWrapper.tsx:245]: Yes, that's right

image
---
image

File: apps/condo/domains/meter/utils/metersImporters/SbbolMetersImporter.ts (hunk 3 above)
Comment (AleX83Xpert) [reply apps/condo/domains/meter/utils/metersImporters/SbbolMetersImporter.ts:35]: Done

File: apps/condo/domains/meter/utils/metersImporters/SbbolMetersImporter.ts (hunk 4 above)
Comment (AleX83Xpert) [reply apps/condo/domains/meter/utils/metersImporters/SbbolMetersImporter.ts:32]: Removed

File: apps/condo/domains/property/utils/testSchema/factories.js (hunk 5 above)
Comment (AleX83Xpert) [reply apps/condo/domains/property/utils/testSchema/factories.js:145]: Done. But this is not the same unitType as `flat`

File: apps/condo/domains/property/utils/testSchema/factories.js (hunk 6 above)
Comment (AleX83Xpert) [reply apps/condo/domains/property/utils/testSchema/factories.js:119]: Done

File: apps/condo/domains/property/utils/testSchema/factories.js (hunk 7 above)
Comment (AleX83Xpert) [reply apps/condo/domains/property/utils/testSchema/factories.js:100]: Looks like it is not necessary to build sections similar to each other. We need just the fact the flat in the building. Not matter which section it in.

File: apps/condo/domains/meter/components/Import/Index.tsx [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+} from '@condo/domains/common/components/Import/Index'
+import { TrackingEventType, useTracking } from '@condo/domains/common/components/TrackingContext'
+import { MetersDataImporter } from '@condo/domains/meter/components/MetersDataImporter'
+import { TOnMetersUpload } from '@condo/domains/meter/components/MetersDataImporterTypes'
+import { useMetersImporter } from '@condo/domains/meter/hooks/useMetersImporter'
+import { ProcessedRow } from '@condo/domains/meter/utils/metersImporters/AbstractMetersImporter'
+
+export type IMetersImportWrapperProps = Pick<IImportWrapperProps, 'accessCheck' | 'onFinish' | 'uploadButtonLabel' | 'importCardButton'>
+
+type ActiveModalType = null | 'example' | 'progress' | 'partlyLoaded' | 'success' | 'error'
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/meter/components/Import/Index.tsx:23]: export/import this type from common import wrapper

File: apps/condo/domains/meter/components/Import/Index.tsx [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+
+const MetersImportWrapper: React.FC<IMetersImportWrapperProps> = (props) => {
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/meter/components/Import/Index.tsx:25]: why can not reuse ImportWrapper from common domain?

File: apps/condo/domains/meter/hooks/useMetersImporter.ts [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+                        [ElectricityResourceTypeValue]: ELECTRICITY_METER_RESOURCE_ID,
+                        [HeatSupplyResourceTypeValue]: HEAT_SUPPLY_METER_RESOURCE_ID,
+                        [GasSupplyResourceTypeValue]: GAS_SUPPLY_METER_RESOURCE_ID,
+                    },
+                }
+            case ImportDataType.sbbol:
+                return {
+                    unitType: {},
+                    resourceId: {
+                        '1': GAS_SUPPLY_METER_RESOURCE_ID, // Газ,
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/meter/hooks/useMetersImporter.ts:164]: Russian comments
It's ok?)

File: apps/condo/domains/meter/utils/metersImporters/AbstractMetersImporter.ts [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+        private mutationErrorsToMessages: MutationErrorsToMessagesType,
+    ) {
+    }
+
+    protected progress = {
+        min: 0,
+        max: 100,
+        current: 0,
+        absProcessed: 0,
+        absTotal: 0,
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/meter/utils/metersImporters/AbstractMetersImporter.ts:50]: what is `abs` prefix?

**Review by nomerdvadcatpyat — COMMENTED** [review]
Do we need to add an example file with new format in meters import modal?

File: apps/condo/domains/meter/schema/RegisterMetersReadingsService.js [hunk 12]
```diff
@@ -58,14 +75,73 @@ const ERRORS = {
...
 
-function toISO (str) {
-    if (isEmpty(str)) {
+const SYMBOLS_TO_CUT_FROM_DATES_REGEXP = /[^-_:.,( )/\d]/g
+
+/**
+ * @param {string} dateStr
+ * @return {undefined|string}
+ */
+function sanitizeDateStr (dateStr) {
```
Comment (dkoviazin) [inline apps/condo/domains/meter/schema/RegisterMetersReadingsService.js:111]: You can move it to utils

File: apps/condo/domains/meter/schema/RegisterMetersReadingsService.js [hunk 13]
```diff
@@ -188,23 +274,55 @@ const RegisterMetersReadingsService = new GQLCustomSchema('RegisterMetersReading
                 const resultRows = []
 
                 for (const reading of readings) {
+                    const meterNumber = reading.meterNumber.trim()
+                    const accountNumber = reading.accountNumber.trim()
```
Comment (dkoviazin) [inline apps/condo/domains/meter/schema/RegisterMetersReadingsService.js:278]: You can use some clear logic from Billing

**Review by dkoviazin — APPROVED** [review]

File: apps/condo/domains/meter/components/Import/Index.tsx (hunk 8 above)
Comment (AleX83Xpert) [reply apps/condo/domains/meter/components/Import/Index.tsx:23]: Done

File: apps/condo/domains/meter/components/Import/Index.tsx (hunk 9 above)
Comment (AleX83Xpert) [reply apps/condo/domains/meter/components/Import/Index.tsx:25]: Refactored

File: apps/condo/domains/meter/hooks/useMetersImporter.ts (hunk 10 above)
Comment (AleX83Xpert) [reply apps/condo/domains/meter/hooks/useMetersImporter.ts:164]: Translated

**Comment by AleX83Xpert** [comment]
> Do we need to add an example file with new format in meters import modal?

No. New format is result of exporting from 1c

**Review by nomerdvadcatpyat — APPROVED** [review]

**Pull request merged**